[Bug]: test_sts_assume_role - a fix is needed

Relevant 4.17 run: 

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/738/33142/1625139/1625234/log

self = <mcg.test_sts_client.TestSTSClient object at 0x7f7ef00e8a30>
mcg_obj_session = <ocs_ci.ocs.resources.mcg.MCG object at 0x7f7f02318df0>
awscli_pod_session = <ocs_ci.ocs.resources.pod.Pod object at 0x7f7f015a7c40>
mcg_account_factory = <function mcg_account_factory_fixture..mcg_account_factory_implementation at 0x7f7efad21f70>
nb_assign_user_role_fixture = <function nb_assign_user_role_fixture..factory at 0x7f7efad218b0>
new_bucket = <function new_bucket..factory at 0x7f7efad21670>
add_env_vars_to_noobaa_core_class = <function add_env_vars_to_noobaa_core_fixture..add_env_vars_to_noobaa_core_implementation at 0x7f7eda71fc10>
add_env_vars_to_noobaa_endpoint_class = <function add_env_vars_to_noobaa_endpoint_fixture..add_env_vars_to_noobaa_endpoint_implementation at 0x7f7eda71f790>


def test_sts_assume_role(
    self,
    mcg_obj_session,
    awscli_pod_session,
    mcg_account_factory,
    nb_assign_user_role_fixture,
    new_bucket,
    add_env_vars_to_noobaa_core_class,
    add_env_vars_to_noobaa_endpoint_class,
):
    """
    Test sts support for Noobaa clients.
    As part of this, we test the following:
        * Assign role to an user
        * Assume the role of an user
        * Perform IO by assuming of the role of another user
        * Remove the role from the user
        * Try to assume the role

    """
    # change sts token expire time to 10 minute
    add_env_vars_to_noobaa_core_class(
        [(constants.STS_DEFAULT_SESSION_TOKEN_EXPIRY_MS, 600000)]
    )
    logger.info("Changed the sts token expiration time to 10 minutes")

    add_env_vars_to_noobaa_endpoint_class(
        [(constants.STS_DEFAULT_SESSION_TOKEN_EXPIRY_MS, 600000)]
    )
    logger.info("Changed the sts token expiration time to 10 minutes")

    # create a bucket using noobaa admin creds
    bucket_1 = "first-bucket"
    retry(ClientError, tries=5, delay=5)(new_bucket)(bucket_name=bucket_1)
    logger.info(f"Created bucket {bucket_1}")

    # create a noobaa account
    user_1 = f"user-{uuid4().hex}"
    nb_account_1 = mcg_account_factory(
        name=user_1,
    )
    signed_request_creds = {
        "region": mcg_obj_session.region,
        "endpoint": mcg_obj_session.sts_internal_endpoint,
        "access_key_id": nb_account_1.get("access_key_id"),
        "access_key": nb_account_1.get("access_key"),
        "ssl": False,
    }

    # create another noobaa account
    user_2 = f"user-{uuid4().hex}"
    nb_account_2 = mcg_account_factory(
        name=user_2,
        allow_bucket_create=False,
    )
    nb_user_access_key_id = nb_account_2.get("access_key_id")
    logger.info(f"Created new user '{user_2}'")

    # assign assume role policy to the user-1
    role_name = "user-1-assume-role"
    nb_assign_user_role_fixture(user_2, role_name, principal=user_1)
    logger.info(f"Assigned the assume role policy to the user {user_2}")

    # noobaa admin assumes the above role

  creds_generated = sts_assume_role(
        awscli_pod_session,
        role_name,
        nb_user_access_key_id,
        signed_request_creds=signed_request_creds,
    )

/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/functional/object/mcg/test_sts_client.py:126: 



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/ocs/bucket_utils.py:2842: in sts_assume_role
    return pod_obj.exec_cmd_on_pod(command=cmd)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/ocs/resources/pod.py:195: in exec_cmd_on_pod
    return self.ocp.exec_oc_cmd(
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/ocs/ocp.py:212: in exec_oc_cmd
    out = run_cmd(
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/utils.py:488: in run_cmd
    completed_process = exec_cmd(



cmd = ['oc', '--kubeconfig', '/home/jenkins/current-cluster-dir/openshift-cluster-dir/auth/kubeconfig', '-n', 's3cli-h0nsfjdaj08ka', 'rsh', ...]
secrets = None, timeout = 600, ignore_error = False, threading_lock = None
silent = False, use_shell = False
cluster_config = <ocs_ci.framework.MultiClusterConfig object at 0x7f7f29361af0>
lock_timeout = 7200, kwargs = {}
_env = {'AUTOMATICALLY_RE_TRIGGER_FAILED_TESTS': 'false', 'BASH_FUNC_which%%': '() {  ( alias;\n eval ${which_declare} ) | /u...read-functions --show-tilde --show-dot $@\n}', 'BUILD_DISPLAY_NAME': '#15897 - j055icm1r3t2', 'BUILD_ID': '15897', ...}
kubeconfig_path = '/home/jenkins/current-cluster-dir/openshift-cluster-dir/auth/kubeconfig'


def exec_cmd(
    cmd,
    secrets=None,
    timeout=600,
    ignore_error=False,
    threading_lock=None,
    silent=False,
    use_shell=False,
    cluster_config=None,
    lock_timeout=7200,
    **kwargs,
):
    """
    Run an arbitrary command locally

    If the command is grep and matching pattern is not found, then this function
    returns "command terminated with exit code 1" in stderr.

    Args:
        cmd (str): command to run
        secrets (list): A list of secrets to be masked with asterisks
            This kwarg is popped in order to not interfere with
            subprocess.run(``**kwargs``)
        timeout (int): Timeout for the command, defaults to 600 seconds.
        ignore_error (bool): True if ignore non zero return code and do not
            raise the exception.
        threading_lock (threading.RLock): threading.RLock object that is used
            for handling concurrent oc commands
        silent (bool): If True will silent errors from the server, default false
        use_shell (bool): If True will pass the cmd without splitting
        cluster_config (MultiClusterConfig): In case of multicluster environment this object
                will be non-null
        lock_timeout (int): maximum timeout to wait for lock to prevent deadlocks (default 2 hours)

    Raises:
        CommandFailed: In case the command execution fails

    Returns:
        (CompletedProcess) A CompletedProcess object of the command that was executed
        CompletedProcess attributes:
        args: The list or str args passed to run().
        returncode (str): The exit code of the process, negative for signals.
        stdout     (str): The standard output (None if not captured).
        stderr     (str): The standard error (None if not captured).

    """
    _env = kwargs.pop("env", os.environ.copy())
    kubeconfig_path = config.RUN.get("kubeconfig")
    if kubeconfig_path:
        _env["KUBECONFIG"] = kubeconfig_path
    if cluster_config:
        kubeconfig_path = cluster_config.RUN.get("kubeconfig")
        if kubeconfig_path:
            _env["KUBECONFIG"] = cluster_config.RUN.get("kubeconfig")
    if isinstance(cmd, str) and not kwargs.get("shell"):
        cmd = shlex.split(cmd)
    if (
        kubeconfig_path
        and cmd[0] == "oc"
        and "--kubeconfig" not in cmd
        and "mirror" not in cmd
    ):
        kube_index = 1
        # check if we have an oc plugin in the command
        plugin_list = "oc plugin list"
        cp = subprocess.run(
            shlex.split(plugin_list),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
        subcmd = cmd[1].split("-")
        if len(subcmd) > 1:
            subcmd = "_".join(subcmd)
        if not isinstance(subcmd, str) and isinstance(subcmd, list):
            subcmd = str(subcmd[0])

        for l in cp.stdout.decode().splitlines():
            if subcmd in l:
                # If oc cmdline has plugin name then we need to push the
                # --kubeconfig to next index
                kube_index = 2
                log.info(f"Found oc plugin {subcmd}")
        cmd = list_insert_at_position(cmd, kube_index, ["--kubeconfig"])
        cmd = list_insert_at_position(cmd, kube_index + 1, [kubeconfig_path])
    try:
        if kwargs.get("shell"):
            masked_cmd = mask_secrets(cmd, secrets)
        else:
            masked_cmd = shlex.join(mask_secrets(cmd, secrets))
        log.info(f"Executing command: {masked_cmd}")
        if threading_lock and cmd[0] == "oc":
            threading_lock.acquire(timeout=lock_timeout)
        completed_process = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdin=subprocess.PIPE,
            timeout=timeout,
            env=_env,
            **kwargs,
        )
    finally:
        if threading_lock and cmd[0] == "oc":
            threading_lock.release()
    masked_stdout = mask_secrets(completed_process.stdout.decode(), secrets)
    if len(completed_process.stdout) > 0:
        log.debug(f"Command stdout: {masked_stdout}")
    else:
        log.debug("Command stdout is empty")

    masked_stderr = mask_secrets(completed_process.stderr.decode(), secrets)
    if len(completed_process.stderr) > 0:
        if not silent:
            log.warning(f"Command stderr: {masked_stderr}")
    else:
        log.debug("Command stderr is empty")
    log.debug(f"Command return code: {completed_process.returncode}")
    if completed_process.returncode and not ignore_error:
        masked_stderr = bin_xml_escape(filter_out_emojis(masked_stderr))
        if (
            "grep" in masked_cmd
            and b"command terminated with exit code 1" in completed_process.stderr
        ):
            log.info(f"No results found for grep command: {masked_cmd}")
        else:

          raise CommandFailed(
                f"Error during execution of command: {masked_cmd}."
                f"\nError is {masked_stderr}"
            )
E               ocs_ci.ocs.exceptions.CommandFailed: Error during execution of command: oc --kubeconfig /home/jenkins/current-cluster-dir/openshift-cluster-dir/auth/kubeconfig -n s3cli-h0nsfjdaj08ka rsh s3cli-0 sh -c 'AWS_ACCESS_KEY_ID=Nh95V7zM86MD5Va5j0ho AWS_SECRET_ACCESS_KEY=goigluzVMvIEm0K4inCRHwR9bAo6ush/QtEmeo49 AWS_DEFAULT_REGION=eu-de aws sts --endpoint=[https://sts.openshift-storage.svc:443](https://sts.openshift-storage.svc/) --no-verify-ssl assume-role --role-arn arn:aws:sts::St8YfQYvQwGxybNQlI3B:role/user-1-assume-role --role-session-name role-session-e030c6b7cd044b618da67b9013861b5d'.
E               Error is /usr/lib/python3.10/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'sts.openshift-storage.svc'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
E                 warnings.warn(
E               /usr/lib/python3.10/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'sts.openshift-storage.svc'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
E                 warnings.warn(
E               /usr/lib/python3.10/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'sts.openshift-storage.svc'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
E                 warnings.warn(
E               /usr/lib/python3.10/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'sts.openshift-storage.svc'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
E                 warnings.warn(
E               /usr/lib/python3.10/site-packages/urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'sts.openshift-storage.svc'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
E                 warnings.warn(
E
E               An error occurred (Unknown) when calling the AssumeRole operation (reached max retries: 4): Unknown
E               command terminated with exit code 255



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/utils.py:720: CommandFailed


2025-06-07 07:14:38

Still relevant - 4.18 occasional failures: 
https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/795/42553/1957761/1958044/log
https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/795/42556/1959057/1959214/log

Backtrace : 
self = <mcg.test_sts_client.TestSTSClient object at 0x7f850c8a2d50>
mcg_obj_session = <ocs_ci.ocs.resources.mcg.MCG object at 0x7f851bfaf490>
awscli_pod_session = <ocs_ci.ocs.resources.pod.Pod object at 0x7f850e450a50>
mcg_account_factory = <function mcg_account_factory_fixture..mcg_account_factory_implementation at 0x7f850f602fc0>
nb_assign_user_role_fixture = <function nb_assign_user_role_fixture..factory at 0x7f850f601e40>
new_bucket = <function new_bucket..factory at 0x7f850f6014e0>
add_env_vars_to_noobaa_core_class = <function add_env_vars_to_noobaa_core_fixture..add_env_vars_to_noobaa_core_implementation at 0x7f850f602480>
add_env_vars_to_noobaa_endpoint_class = <function add_env_vars_to_noobaa_endpoint_fixture..add_env_vars_to_noobaa_endpoint_implementation at 0x7f850f601da0>


def test_sts_assume_role(
    self,
    mcg_obj_session,
    awscli_pod_session,
    mcg_account_factory,
    nb_assign_user_role_fixture,
    new_bucket,
    add_env_vars_to_noobaa_core_class,
    add_env_vars_to_noobaa_endpoint_class,
):
    """
    Test sts support for Noobaa clients.
    As part of this, we test the following:
        * Assign role to an user
        * Assume the role of an user
        * Perform IO by assuming of the role of another user
        * Remove the role from the user
        * Try to assume the role

    """
    # change sts token expire time to 10 minute
    add_env_vars_to_noobaa_core_class(
        [(constants.STS_DEFAULT_SESSION_TOKEN_EXPIRY_MS, 600000)]
    )
    logger.info("Changed the sts token expiration time to 10 minutes")

    add_env_vars_to_noobaa_endpoint_class(
        [(constants.STS_DEFAULT_SESSION_TOKEN_EXPIRY_MS, 600000)]
    )
    logger.info("Changed the sts token expiration time to 10 minutes")

    # create a bucket using noobaa admin creds
    bucket_1 = "first-bucket"

  retry(ClientError, tries=5, delay=5)(new_bucket)(bucket_name=bucket_1)

/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/functional/object/mcg/test_sts_client.py:95: 



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/utility/retry.py:99: in f_retry
    return f(args, *kwargs)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/tests/functional/object/mcg/test_sts_client.py:33: in factory
    s3_create_bucket(mcg_obj_session, bucket_name, s3_client)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/ocs_ci/ocs/bucket_utils.py:2879: in s3_create_bucket
    s3_obj.s3_resource.create_bucket(Bucket=bucket_name)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/venv/lib64/python3.11/site-packages/boto3/resources/factory.py:581: in do_action
    response = action(self, args, *kwargs)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/venv/lib64/python3.11/site-packages/boto3/resources/action.py:88: in call
    response = getattr(parent.meta.client, operation_name)(args, *params)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/venv/lib64/python3.11/site-packages/botocore/client.py:595: in _api_call
    return self._make_api_call(operation_name, kwargs)
/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/venv/lib64/python3.11/site-packages/botocore/context.py:123: in wrapper
    return func(args, *kwargs)



self = <botocore.client.S3 object at 0x7f851c2da0d0>
operation_name = 'CreateBucket', api_params = {'Bucket': 'first-bucket'}


@with_current_context()
def _make_api_call(self, operation_name, api_params):
    operation_model = self._service_model.operation_model(operation_name)
    service_name = self._service_model.service_name
    history_recorder.record(
        'API_CALL',
        {
            'service': service_name,
            'operation': operation_name,
            'params': api_params,
        },
    )
    if operation_model.deprecated:
        logger.debug(
            'Warning: %s.%s() is deprecated', service_name, operation_name
        )
    request_context = {
        'client_region': self.meta.region_name,
        'client_config': self.meta.config,
        'has_streaming_input': operation_model.has_streaming_input,
        'auth_type': operation_model.resolved_auth_type,
        'unsigned_payload': operation_model.unsigned_payload,
    }

    api_params = self._emit_api_params(
        api_params=api_params,
        operation_model=operation_model,
        context=request_context,
    )
    (
        endpoint_url,
        additional_headers,
        properties,
    ) = self._resolve_endpoint_ruleset(
        operation_model, api_params, request_context
    )
    if properties:
        # Pass arbitrary endpoint info with the Request
        # for use during construction.
        request_context['endpoint_properties'] = properties
    request_dict = self._convert_to_request_dict(
        api_params=api_params,
        operation_model=operation_model,
        endpoint_url=endpoint_url,
        context=request_context,
        headers=additional_headers,
    )
    resolve_checksum_context(request_dict, operation_model, api_params)

    service_id = self._service_model.service_id.hyphenize()
    handler, event_response = self.meta.events.emit_until_response(
        f'before-call.{service_id}.{operation_name}',
        model=operation_model,
        params=request_dict,
        request_signer=self._request_signer,
        context=request_context,
    )

    if event_response is not None:
        http, parsed_response = event_response
    else:
        maybe_compress_request(
            self.meta.config, request_dict, operation_model
        )
        apply_request_checksum(request_dict)
        http, parsed_response = self._make_request(
            operation_model, request_dict, request_context
        )

    self.meta.events.emit(
        f'after-call.{service_id}.{operation_name}',
        http_response=http,
        parsed=parsed_response,
        model=operation_model,
        context=request_context,
    )

    if http.status_code >= 300:
        error_info = parsed_response.get("Error", {})
        error_code = error_info.get("QueryErrorCode") or error_info.get(
            "Code"
        )
        error_class = self.exceptions.from_code(error_code)

      raise error_class(parsed_response, operation_name)
E           botocore.errorfactory.BucketAlreadyOwnedByYou: An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it. You get this error in all AWS regions except US East (N. Virginia) region, us-east-1. In us-east-1 region, you will get 200 OK, but it is no-op (if bucket exists it Amazon S3 will not do anything).



/home/jenkins/workspace/qe-deploy-ocs-cluster-prod/ocs-ci/venv/lib64/python3.11/site-packages/botocore/client.py:1058: BucketAlreadyOwnedByYou

--- a/tests/functional/object/mcg/test_sts_client.py
+++ b/tests/functional/object/mcg/test_sts_client.py
@@ -91,14 +91,8 @@ class TestSTSClient:
         logger.info("Changed the sts token expiration time to 10 minutes")
 
         # create a bucket using noobaa admin creds
-        bucket_1 = "first-bucket"
-        try:
-            retry(ClientError, tries=5, delay=5)(new_bucket)(bucket_name=bucket_1)
-        except ClientError as e:
-            if "BucketAlreadyExists" in str(e):
-                logger.info(f"Bucket {bucket_1} already exists")
-            else:
-                raise
+        bucket_1 = f"first-bucket-{uuid4().hex[:8]}"
+        retry(ClientError, tries=5, delay=5)(new_bucket)(bucket_name=bucket_1)
         logger.info(f"Created bucket {bucket_1}")
 
         # create a noobaa account
@@ -129,7 +123,7 @@ class TestSTSClient:
         logger.info(f"Assigned the assume role policy to the user {user_2}")
 
         # noobaa admin assumes the above role
-        creds_generated = sts_assume_role(
+        creds_generated = retry(CommandFailed, tries=5, delay=30)(sts_assume_role)(
             awscli_pod_session,
             role_name,
             nb_user_access_key_id,
@@ -140,7 +134,7 @@ class TestSTSClient:
         )
 
         # perform io to validate the role assumption
-        bucket_2 = "second-bucket"
+        bucket_2 = f"second-bucket-{uuid4().hex[:8]}"
         try:
             new_bucket(bucket_2, assumed_user_s3client)
             assert (

